### PR TITLE
Trip tests

### PR DESCRIPTION
**Pull request by RCAVelez**

### CODE REVIEW

File: backend/src/test/java/com/google/sps/TripCRUDTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  @Test
+  public void readTrip_returnsEntityForTrip() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCRUD.createTrip(EMAIL, TRIP_DATA);
+    Entity readTripEntity = new Entity("Trip");
+    try {
+      readTripEntity = TripCRUD.readTrip(Long.toString(tripEntity.getKey().getId()));
+    } catch (EntityNotFoundException e) {
+    }
```
Comment (Cindyzhang977) [inline backend/src/test/java/com/google/sps/TripCRUDTest.java:73]: Instead of try/catch in a junit test, you can explicitly state that an error is expected. I think something like `@Test(expected = EntityNotFoundException.class)` should work (might need some imports?). This would require you to split this test in two: one where an error is expected to be thrown and one where you `asserEquals`. [(Ref)](https://www.guru99.com/junit-exception-test.html#:~:text=JUnit%20provides%20the%20facility%20to,throwing%20expected%20exception%20or%20not.&text=While%20Testing%20exception%2C%20you%20need,test%20annotation%20is%20the%20same.)

File: backend/src/test/java/com/google/sps/TripCRUDTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  @Test
+  public void updateTrip_returnsUpdatedTripNameForUpdatedEntity() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCRUD.createTrip(EMAIL, TRIP_DATA);
+    TripCRUD.updateTrip(Long.toString(tripEntity.getKey().getId()), TRIP_DATA_2);
+    Entity tripFound = new Entity("Trip");
+    try {
+      tripFound = TripCRUD.readTrip(Long.toString(tripEntity.getKey().getId()));
+    } catch (EntityNotFoundException e) {
+    }
```
Comment (Cindyzhang977) [inline backend/src/test/java/com/google/sps/TripCRUDTest.java:100]: same thing for the try/catch here

File: .gitignore [hunk 3]
```diff
@@ -191,6 +191,9 @@ dist
 **/*.backup.*
```
Comment (mecolmg) [inline .gitignore:191]: Can you fix your git branch to not show these and other unrelated changes?

File: backend/src/test/java/com/google/sps/TripCRUDTest.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import static org.junit.Assert.assertEquals;
+
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCRUDTest.java:20]: nit: unused

File: backend/src/main/java/com/google/sps/TripCRUD.java [hunk 5]
```diff
@@ -84,9 +84,15 @@ private static void setProperties(Entity tripEntity, String tripData) {
...
-  public static Entity readTrip(String tripId) throws EntityNotFoundException {
+  public static Entity readTrip(String tripId) {
     Key entityKey = KeyFactory.createKey("Trip", Long.parseLong(tripId));
-    return datastore.get(entityKey);
+    Entity tripEntity;
+    try {
+      tripEntity = datastore.get(entityKey);
+    } catch (EntityNotFoundException e) {
+      return null;
+    }
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCRUD.java:94]: Did you change this for testing? In general, you shouldn't change functionality of what you're testing to make it easier to test. That's a codesmell.

To test exceptions in JUnit4, see: https://www.baeldung.com/junit-assert-exception#junit-4

File: backend/src/main/java/com/google/sps/servlets/readTripServlet.java [hunk 6]
```diff
@@ -35,12 +35,11 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     }
 
-    Entity tripEntity;
-    try {
-      tripEntity = TripCRUD.readTrip(tripId);
-    } catch (EntityNotFoundException e) {
-      return;
+    Entity tripEntity = TripCRUD.readTrip(tripId);
+    if (tripEntity == null) {
+      throw new IllegalArgumentException("trip not found");
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/servlets/readTripServlet.java:40]: nit:
```suggestion
      throw new IllegalArgumentException("Trip not found");
```

File: backend/src/test/java/com/google/sps/TripCRUDTest.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCRUD.createTrip(EMAIL, TRIP_DATA);
+    Entity actual = TripCRUD.readTrip(INVALID_TRIP_ID);
+    assertEquals(null, actual);
+  }
+
+  @Test
+  public void toEntity_returnsTripEntityFromJsonMatchingTripName() {
+    Entity tripEntityConverted = TripCRUD.toEntity(TRIP_DATA, "", null);
+    assertEquals("\"My Milan Trip\"", (String) tripEntityConverted.getProperty("tripName"));
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCRUDTest.java:79]: Hmm, should it have the extra quotes stored? Is this a bug?

File: backend/src/test/java/com/google/sps/TripCRUDTest.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    Entity tripEntity = TripCRUD.createTrip(EMAIL, TRIP_DATA);
+    TripCRUD.updateTrip(Long.toString(tripEntity.getKey().getId()), TRIP_DATA_2);
+    Entity tripFound = TripCRUD.readTrip(Long.toString(tripEntity.getKey().getId()));
+    assertEquals("\"My Awesome Milan Trip\"", (String) tripFound.getProperty("tripName"));
+  }
+
+  @Test
+  public void updateTrip_noTripFound() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCRUD.createTrip(EMAIL, TRIP_DATA);
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCRUDTest.java:102]: nit: not really necessary

File: backend/src/test/java/com/google/sps/TripCRUDTest.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    Entity actual = TripCRUD.readTrip(INVALID_TRIP_ID);
+    assertEquals(null, actual);
+  }
+}
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCRUDTest.java:106]: nit: add edge case tests for:
- Creating a trip without creating a user first
- toJson with missing entity info
- toEntity with missing or malformed json
- Any other edge cases you can think of.

File: frontend/src/RouteView.js [hunk 10]
```diff
@@ -60,6 +61,13 @@ function RouteView({ loggedIn }) {
   function save() {
     setIsSaved(true);
     // save to back end database
+    tripObject.isOptimized = isOptimized;
+    console.log(tripObject);
+    if (!tripObject.id) {
+      createTrip(userEmail, tripObject).then((tripId) => console.log(tripId));
+    } else {
+      updateTrip(tripObject.id, tripObject);
+    }
```
Comment (Cindyzhang977) [inline frontend/src/RouteView.js:70]: please ignore this. it was accidentally pushed from another branch when trying to push java updates.

File: backend/src/test/java/com/google/sps/TripCRUDTest.java (hunk 9 above)
Comment (zoebarth) [reply backend/src/test/java/com/google/sps/TripCRUDTest.java]: Done

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 11]
```diff
@@ -27,8 +26,21 @@
...
+  /**
+   * Add tripId property to Trip Entity
+   *
+   * @param id the trip id
+   */
+  public static void addTripId(String id) {
+    Entity tripEntity = TripCrud.readTrip(id);
+    tripEntity.setProperty("tripId", id);
+    datastore.put(tripEntity);
   }
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:44]: You shouldn't need to store tripId as a property of the entity, since you can get it as entity.getKey().getId()

I think this method can be removed, and if for some reason it can't, it should be made private.

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 12]
```diff
@@ -55,22 +67,21 @@ private static void setProperties(Entity tripEntity, String tripData) {
...
+      embeddedAttraction.setProperty("photoUrl", attraction.get("photoUrl").toString());
       embeddedAttraction.setProperty(
-          "attractionName", attraction.get("attractionName").getAsString());
-      embeddedAttraction.setProperty(
-          "photoReference", attraction.get("photoReference").getAsString());
-      embeddedAttraction.setProperty("routeIndex", attraction.get("routeIndex").getAsInt());
-      embeddedAttraction.setProperty("coordinates", attraction.get("coordinates").getAsString());
+          "routeIndex", Integer.parseInt(attraction.get("routeIndex").toString()));
+      embeddedAttraction.setProperty("lat", attraction.get("lat").toString());
+      embeddedAttraction.setProperty("lng", attraction.get("lng").toString());
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:84]: nit: This seems like a regression, what was wrong with the existing code?

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 13]
```diff
@@ -27,8 +26,21 @@
   public static Entity createTrip(String email, String tripData) {
     Entity tripEntity = toEntity(tripData, "", null);
     datastore.put(tripEntity);
-    UserCrud.addTripId(email, tripEntity.getKey().getId());
-    return tripEntity;
+    long id = tripEntity.getKey().getId();
+    UserCrud.addTripId(email, id);
+    TripCrud.addTripId(Long.toString(id));
+    return TripCrud.readTrip(Long.toString(id));
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:32]: Why can't you just return tripEntity?

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 14]
```diff
@@ -80,12 +91,17 @@ private static void setProperties(Entity tripEntity, String tripData) {
    * Finds a trip entity by id
    *
    * @param tripId id of the trip to find
-   * @throws EntityNotFoundException if trip entity is not found
-   * @return Trip entity
+   * @return Trip entity if found, null if EntityNotFoundException is caught
    */
-  public static Entity readTrip(String tripId) throws EntityNotFoundException {
+  public static Entity readTrip(String tripId) {
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:96]: `entity.getKey().getId()` returns a long, so it would make more sense for this to take in a long.

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 15]
```diff
@@ -97,23 +113,24 @@ public static Entity readTrip(String tripId) throws EntityNotFoundException {
...
-          "photoReference", (String) attraction.getProperty("photoReference"));
-      attractionJson.addProperty("routeIndex", (Integer) attraction.getProperty("routeIndex"));
-      attractionJson.addProperty("coordinates", (String) attraction.getProperty("coordinates"));
+          "routeIndex", Integer.parseInt(attraction.getProperty("routeIndex").toString()));
+      attractionJson.addProperty("lat", attraction.getProperty("lat").toString());
+      attractionJson.addProperty("lng", attraction.getProperty("lng").toString());
       attractions.add(attractionJson);
     }
-    jsonTrip.add("attractions", attractions);
+    jsonTrip.addProperty("attractions", attractions.toString());
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:133]: Likewise, this seems like a regression. Why are you storing this as a string? This will mess with deserializing the object from the client.

The way it was implemented previously by storing a JsonArray using the `add` method seems like the correct way to do this: https://www.javadoc.io/static/com.google.code.gson/gson/2.6.2/com/google/gson/JsonObject.html#add-java.lang.String-com.google.gson.JsonElement-

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 16]
```diff
@@ -124,10 +141,8 @@ public static JsonObject toJson(Entity tripEntity) {
    * @param tripData string representation of tripData json
    */
   public static void updateTrip(String tripId, String tripData) {
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:143]: Likewise, tripId should probably be a long

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    assertEquals(tripEntity, readTripEntity);
+  }
+
+  @Test
+  public void readTrip_returnsEntityForTrip() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCrud.createTrip(EMAIL, TRIP_DATA);
+    Entity readTripEntity = TripCrud.readTrip(Long.toString(tripEntity.getKey().getId()));
+    assertEquals(tripEntity, readTripEntity);
+  }
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:66]: These two tests are the same

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+  public void tearDown() {
+    DATASTORE_SERVICE_HELPER.tearDown();
+  }
+
+  @Test
+  public void createTrip_returnsEntityForCreatedTrip() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCrud.createTrip(EMAIL, TRIP_DATA);
+    Entity readTripEntity = TripCrud.readTrip(Long.toString(tripEntity.getKey().getId()));
+    assertEquals(tripEntity, readTripEntity);
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:57]: Instead of this, you should create an expected Entity and verify that it equals the actual entity, or you could verify that each of the properties is instantiated as expected.

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  @Test
+  public void readTrip_noTripFound() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCrud.createTrip(EMAIL, TRIP_DATA);
+    Entity actual = TripCrud.readTrip(INVALID_TRIP_ID);
+    assertEquals(null, actual);
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:73]: nit: use assertNull

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 20]
```diff
@@ -27,8 +26,21 @@
...
+   *
+   * @param id the trip id
+   */
+  public static void addTripId(String id) {
+    Entity tripEntity = TripCrud.readTrip(id);
+    tripEntity.setProperty("tripId", id);
+    datastore.put(tripEntity);
   }
 
   /**
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:46]: Doesn't need to be in this PR, but there should really be two toEntity methods. One that takes in tripData, and one that takes in the tripData, tripId, and tripKey. Having to pass in null or empty strings to a method is a **big** codesmell in Java.

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  @Test
+  public void toEntity_returnsTripEntityFromJsonMatchingTripName() {
+    Entity tripEntityConverted = TripCrud.toEntity(TRIP_DATA, "", null);
+    assertEquals("\"My Milan Trip\"", (String) tripEntityConverted.getProperty("tripName"));
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:79]: What's up with the extra quotes? I wouldn't expect those to be stored.

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  @Test
+  public void toJson_returnsTripJsonFromEntityMatchingSearchText() {
+    Entity userEntity = UserCrud.createUser(EMAIL);
+    Entity tripEntity = TripCrud.createTrip(EMAIL, TRIP_DATA);
+    JsonObject tripDataJson = TripCrud.toJson(tripEntity);
+    assertEquals("\"Milano\"", tripDataJson.get("searchText").getAsString());
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:87]: Same with extra quotes

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 21 above)
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:79]: Can we check more than just the trip name? Testing the entire entity would be preferable, but you should at least test the attractions.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 22 above)
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:87]: Same here with testing the entire expected JSON or at least the attractions.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 18 above)
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:57]: nit: for readability separate your actions and expectations by a newline here and below.

```suggestion
    Entity userEntity = UserCrud.createUser(EMAIL);
    Entity tripEntity = TripCrud.createTrip(EMAIL, TRIP_DATA);
    Entity readTripEntity = TripCrud.readTrip(Long.toString(tripEntity.getKey().getId()));

    assertEquals(tripEntity, readTripEntity);
```

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+    Entity tripEntity = TripCrud.createTrip(EMAIL, TRIP_DATA);
+    TripCrud.updateTrip(Long.toString(tripEntity.getKey().getId()), TRIP_DATA_2);
+    Entity tripFound = TripCrud.readTrip(Long.toString(tripEntity.getKey().getId()));
+    assertEquals("\"My Awesome Milan Trip\"", (String) tripFound.getProperty("tripName"));
+  }
+
+  @Test
+  public void updateTrip_noTripFound() {
+    Entity actual = TripCrud.readTrip(INVALID_TRIP_ID);
+    assertEquals(null, actual);
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:102]: nit: assertNull

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  @Test(expected = Exception.class)
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:105]: nit: Is there a more specific exception you can verify throws than `Exception`? The reason being, if something else breaks in createTrip, this test will still pass. Same below

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 12 above)
Comment (Cindyzhang977) [reply backend/src/main/java/com/google/sps/TripCrud.java:84]: We changed the way the json file was represented a while ago, so we had to update this to match.

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 11 above)
Comment (Cindyzhang977) [reply backend/src/main/java/com/google/sps/TripCrud.java:44]: This is added to send the json to the front end. The tripId is used to get the trip from the database and if it's null it indicates the trip must be created.

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 13 above)
Comment (Cindyzhang977) [reply backend/src/main/java/com/google/sps/TripCrud.java:32]: `tripEntity` was updated in `TripCrud.addTripId` so we should read it again from the database to get the most up to date version.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 23 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:102]: done

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 19 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:73]: done

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 17 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:66]: I removed the readTrip_returnsEntityForTrip test

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 18 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:57]: done, I added a space between the assert and the logic above it on all tests

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 24 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:105]: done. All expected outcomes now have the a specific exception. 
These type of tests seem hacky. Are these fine?

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 11 above)
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/TripCrud.java:44]: done, it's private

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 14 above)
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/TripCrud.java:96]: readTrip, updateTrip, and toEntity all now take in a tripId of type Long. The servlets and tests have now been changed to reflect that change in type.

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 16 above)
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/TripCrud.java:143]: Done! updateTrip now takes in a tripId of type Long. The servlets and tests have now been changed to reflect that change in type.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 21 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:79]: We are having an issue with the toEntity methods where the additional quote is added. In order to pass this test we have to add an additional quote. Then in the frontend we have to remove that additional quote.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 22 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:87]: We are having an issue with the toJson method where the additional quote is added. In order to pass this test we have to add an additional quote. Then in the frontend we have to remove that additional quote.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 21 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:79]: yes we can

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 22 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:87]: can do

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 20 above)
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/TripCrud.java:46]: Could we convert this to an issue and resolve it Monday?

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 15 above)
Comment (RCAVelez) [reply backend/src/main/java/com/google/sps/TripCrud.java:131]: should be fixed now

File: backend/src/main/java/com/google/sps/TripCrud.java (hunk 20 above)
Comment (mecolmg) [reply backend/src/main/java/com/google/sps/TripCrud.java:46]: Sure.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 21 above)
Comment (mecolmg) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:75]: I guess that's my point. That means something is wrong, that needs to be fixed. The current workaround seems very hacky.

Comment (mecolmg) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:75]: This can be converted to an issue and solved in a later PR

Comment (mecolmg) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:75]: I don't see any changes though

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 22 above)
Comment (mecolmg) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:84]: I also don't see this in the changes.

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 21 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java:75]: done

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 22 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java]: done, I created a test for every property of attraction, and I was able to remove the extra quotes

File: backend/src/test/java/com/google/sps/TripCrudTest.java (hunk 21 above)
Comment (RCAVelez) [reply backend/src/test/java/com/google/sps/TripCrudTest.java]: I was able to remove the extra quotes

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 25]
```diff
@@ -68,18 +69,23 @@ private static void setProperties(Entity tripEntity, String tripData) {
...
-    tripEntity.setProperty("searchText", jsonObject.get("searchText").toString());
-    tripEntity.setProperty("tripName", jsonObject.get("tripName").toString());
-    tripEntity.setProperty("centerLocation", jsonObject.get("centerLocation").toString());
+    tripEntity.setProperty("searchText", jsonObject.get("searchText").getAsString());
+    tripEntity.setProperty("tripName", jsonObject.get("tripName").getAsString());
+
+    Integer centerLat = jsonObject.get("centerLat").getAsInt();
+    Integer centerLng = jsonObject.get("centerLng").getAsInt();
+    tripEntity.setProperty("centerLng", centerLng);
+    tripEntity.setProperty("centerLat", centerLat);
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:78]: optional: You could make the centerLocation an `EmbededEntity` that has `lat` and `lng`.

File: backend/src/main/java/com/google/sps/TripCrud.java [hunk 26]
```diff
@@ -68,18 +69,23 @@ private static void setProperties(Entity tripEntity, String tripData) {
     JsonElement jsonElement = parser.parse(tripData);
     JsonObject jsonObject = jsonElement.getAsJsonObject();
     tripEntity.setProperty("isOptimized", jsonObject.get("isOptimized").getAsBoolean());
-    tripEntity.setProperty("searchText", jsonObject.get("searchText").toString());
-    tripEntity.setProperty("tripName", jsonObject.get("tripName").toString());
-    tripEntity.setProperty("centerLocation", jsonObject.get("centerLocation").toString());
+    tripEntity.setProperty("searchText", jsonObject.get("searchText").getAsString());
```
Comment (mecolmg) [inline backend/src/main/java/com/google/sps/TripCrud.java:72]: Ah, nice catch! I'm glad you caught this. FYI @Cindyzhang977 you may need to adjust for this fix in your dependent PRs

File: backend/src/test/java/com/google/sps/TripCrudTest.java [hunk 27]
```diff
@@ -72,7 +76,80 @@ public void readTrip_noTripFound() {
   public void toEntity_returnsTripEntityFromJsonMatchingTripName() {
     Entity tripEntityConverted = TripCrud.toEntity(TRIP_DATA, null, null);
 
-    assertEquals("\"My Milan Trip\"", (String) tripEntityConverted.getProperty("tripName"));
+    assertEquals("My Milan Trip", (String) tripEntityConverted.getProperty("tripName"));
```
Comment (mecolmg) [inline backend/src/test/java/com/google/sps/TripCrudTest.java:79]: optional: All these assertions could be done in a single test. It's fine to have multiple assertions in a single test, but they should generally be asserting the same thing. Here, you're verifying that toEntity converts from a json object to an entity.
